src/plenoptic/simulate/models/portilla_simoncelli.py

```diff
@@ -82,934 +92,849 @@ def __init__(
+            # Then compute the outer product of those variances.
+            var_outer_prod = einops.einsum(coeff_var, coeff_other_var,
+                                           'b c o1, b c o2 -> b c o1 o2')
+            # And the sqrt of this is what we use to normalize the covariance
+            # into the cross-correlation
+            covars.append(covar / var_outer_prod.pow(.5))
+        return torch.stack(covars, -1), torch.stack(coeffs_var, -1)
+
+    def _double_phase_pyr_coeffs(self, pyr_coeffs: List[Tensor]) -> Tuple[List[Tensor], List[Tensor]]:
+        """Upsample and double the phase of pyramid coefficients.
 
         Parameters
         ----------
-        X: torch.Tensor
-            matrix to compute the kurtosis of.
-        mu: torch.Tensor
-            pre-computed mean. If None, we compute it.
-        var: torch.Tensor
-            pre-computed variance. If None, we compute it.
+        pyr_coeffs :
+            Complex steerable pyramid coefficients (without residuals), as list
+            of length n_scales, containing 5d tensors of shape (batch, channel,
+            n_orientations, height, width)
 
         Returns
         -------
-        kurtosis: torch.Tensor
-            kurtosis of the matrix X
+        doubled_phase_mags :
+            The demeaned magnitude (i.e., pyr_coeffs.abs()) of each upsampled
+            double-phased coefficient. List of length n_scales-1 containing
+            tensors of same shape the input (the finest scale has been
+            removed).
+        doubled_phase_separate :
+            The real and imaginary parts of each double-phased coefficient.
+            List of length n_scales-1, containing tensors of shape (batch,
+            channel, 2*n_orientations, height, width), with the real component
+            found at the same orientation index as the input, and the imaginary
+            at orientation+self.n_orientations. (The finest scale has been
+            removed.)
 
         """
-        # implementation is only for real components
-        if mu is None:
-            mu = X.mean()
-        if var is None:
-            var = X.var()
-        return torch.mean(torch.abs(X - mu).pow(4)) / (var.pow(2))
+        doubled_phase_mags = []
+        doubled_phase_sep = []
+        # don't do this for the finest scale
+        for coeff in pyr_coeffs[1:]:
+            doubled_phase = signal.expand(coeff, 2) / 4.0
+            doubled_phase = signal.modulate_phase(doubled_phase, 2)
+            doubled_phase_mag = doubled_phase.abs()
+            doubled_phase_mag = doubled_phase_mag - doubled_phase_mag.mean((-2, -1), keepdim=True)
+            doubled_phase_mags.append(doubled_phase_mag)
+            doubled_phase_sep.append(einops.pack([doubled_phase.real, doubled_phase.imag],
+                                                 'b c * h w')[0])
+        return doubled_phase_mags, doubled_phase_sep
 
+    def plot_representation(
```

Comment:
yeah, I don't love how I'm handling plotting right now. the idea is that plotting for every model will be different and the way to combine from the complete representation into a format for plotting is model-specific and not obvious. This gets called in the synthesis plotting code (which you can see in the PortillaSimoncelli notebook or the Metamer notebook): we just call `model.plot_representation()` and, if that method doesn't exist, use some default (single stem plot if the representation looks like a vector, image plot if it looks like an image). I didn't want to try and get more any clever.

But I'm not super happy with how I handle plotting, because it is messy, like you said. I'd be happy to talk about this more. But this might end up being a separate PR.

Reply:
yep, I agree with the separate PR